ramips: add support for Netgear R6080

alexclewontin opened the pull request:
This adds support for the Netgear R6080, aka Netgear AC1000. 

The R6080 has almost the same hardware as the Netgear R6120,
aka Netgear AC1200. Significant adaptations:
- Removed kernel module packages providing USB support
- Remapped MTD partitions to fit within 8 MiB, instead of 16 MiB.
- Used correct SERCOMM_HWID in image header, which otherwise would
cause rejection of the image.

Specification:
- SoC: MediaTek MT7628 (580 MHz)
- Flash: 8 MiB
- RAM: 64 MiB
- Wireless: 2.4Ghz (builtin) and 5Ghz (MT7612E)
- LAN speed: 10/100
- LAN ports: 4
- WAN speed: 10/100
- WAN ports: 1
- UART (57600 8N1) on PCB

For initial flash, use nmrpflash with the resulting factory.img file, as
detailed on the R6120 wiki page. Once OpenWRT has been installed,
subsequent flashes can use the web interface and sysupgrade files.

Signed-off-by: Alex Lewontin <alex.c.lewontin@gmail.com>

adrianschmutzler commented:
Please add the description to the commit, not just to the PR.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,147 @@
+			gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan {
```
that would be wlan2g_green.

No 5G led?

alexclewontin replied:
No 5G LED

Four LEDs:
- Power (green only)
- WLAN 2G (green/orange)
- WAN (green/orange)
- LAN (green only)

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,147 @@
+			gpios = <&gpio1 10 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan_orange {
```
wlan2g_orange

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,147 @@
+			gpios = <&gpio1 9 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
```
wan_green

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,147 @@
+	gpio {
+		ralink,group = "p0led_an", "p1led_an", "p2led_an",
+			       "p3led_an", "p4led_an", "wdt",
+			       "wled_an", "uart1";
```
Why uart1?

adrianschmutzler replied:
Ah, it was for the USB power you removed. So, you can remove the uart1 as well.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,147 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Does this support higher speed?

alexclewontin replied:
The data sheet for the chip doesn't give a definitive answer (different values for different scenarios, some of which work and some of which do not). Trial-and-error suggests that it works up to a max value of 64000000.

dengqf6 replied:
Use `m25p,fast-read;` property to get a higher frequency such as 80000000

alexclewontin replied:
Funny you should suggest that, am playing with that right now. That lets it support up to 80000000, but anecdotally, the initial boot with FAST_READ enabled at 80 MHz takes almost twice as long as when its only set to 64 MHz. Should I try for the highest max freq regardless of time, or try to optimize performance?

alexclewontin replied:
Initial boot with FAST_READ enabled at 80 MHz - ~140 seconds
Initial boot with FAST_READ enabled at 64 MHz - ~64 seconds
Initial boot with FAST_READ disabled at 64 MHz - ~78 seconds

dengqf6 replied:
Have you actually tested read speed?
`time cat /dev/mtd3 > /dev/null`

alexclewontin replied:
Ah, that definitely shows that higher is better. The data sheet specifies up to 86 MHz, and that works. Thanks!

alexclewontin commented:
@adrianschmutzler A lot of these hardware names are directly from the mt7628an_netgear_r6120.dts file in the same location. Since they need to change and the R6080 and R6120 overlap significantly, worth consolidating into a mt7628an_netgear.dtsi file? Happy to do that as I address things that need fixing.

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds:
```diff
@@ -40,6 +40,7 @@ hiwifi,hc5761a)
 mediatek,linkit-smart-7688)
 	ucidef_set_led_wlan "wifi" "wifi" "linkit-smart-7688:orange:wifi" "phy0tpt"
 	;;
+netgear,r6080|\
```
Please add a DTS trigger for the wifi instead of specifying it here.

adrianschmutzler commented:
> worth consolidating into a mt7628an_netgear.dtsi file

If the partition layout is different, I wouldn't do that for now.

> The R6080 has almost the same hardware as the Netgear R6120,
aka Netgear AC1200. Significant adaptations:
> Removed kernel module packages providing USB support
> Remapped MTD partitions to fit within 8 MiB, instead of 16 MiB.
> Used correct SERCOMM_HWID in image header, which otherwise would
> cause rejection of the image.

"The R6080 has almost the same hardware as the Netgear R6120,
aka Netgear AC1200, but it lacks the USB port, has only 8 MiB flash and uses different SERCOMM_HWID." or similar

Have you verified the partition layout with stock firmware? Is there really a "reserved" partition followed by "reserved1".

If no usb port is there, you should add "status=disabled" for &ehci and &ohci.

adrianschmutzler commented:
Ah, and are there other valid looking MAC addresses in factory except at 0x4?

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -90,6 +90,7 @@ ramips_setup_interfaces()
+	netgear,r6080|\
 	netgear,r6120)
 		ucidef_add_switch "switch0" \
 			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "6@eth0"
```
Have you checked that the LAN port numbering on the case actually is inverted here as well?

alexclewontin replied:
Yes, double checked this and it is inverted.

alexclewontin commented:
@adrianschmutzler Everything has been addressed, with the exception of the MAC addresses. 

The MAC address in factory at 0x4 (has a Netgear OUI, e.g. B0:39:56:00:00:00) is what is listed on the case and seems to be used for most interfaces (br-lan, eth0, eth0.1, and wlan0). However, eth0.2 uses the MAC address immediately above that one (e.g. B0:39:56:00:00:01, not listed anywhere in the factory partition) and wlan1 uses the MAC address two above it (e.g. B0:39:56:00:00:02, which appears in the factory partition at 0x8044). 

There also appear to be what could be two consecutive Ralink/MediaTek MAC addresses at 0x28 and 0x2E, but I haven't seen those appear anywhere else, and am not quite sure what to make of them.

adrianschmutzler commented:
> There also appear to be what could be two consecutive Ralink/MediaTek MAC addresses at 0x28 and 0x2E, but I haven't seen those appear anywhere else, and am not quite sure what to make of them.

These might be unused dummy addresses. If they are not used in vendor firmware, don't use them either.

> wlan1 uses the MAC address two above it (e.g. B0:39:56:00:00:02, which appears in the factory partition at 0x8044).

Should be 0x8004? If you have a valid address there that matches 5 GHz MAC address on stock FW, you can remove the
```
mtd-mac-address = <&factory 0x4>;
mtd-mac-address-increment = <(2)>;
```
from wifi@0,0. (If mtd-eeprom is 0x8000, the MAC address will be read from 0x8004 automatically.)

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,154 @@
+		lan {
+			label = "r6080:green:lan";
+			gpios = <&gpio1 12 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "";
```
Don't think this has any use? (same below)

alexclewontin replied:
Yes, sorry, was considering adding a trigger, but forgot to remove

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,154 @@
+};
+
+&ehci {
+
```
Remove empty line.

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6120.dts:
```diff
@@ -40,17 +40,18 @@
+		wlan2g_green {
 			label = "r6120:green:wlan2g";
 			gpios = <&gpio1 10 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
```
Please do not mix changes to r6120 with adding R6080.

Have you tested these changes on R6120?

Please separate changes by creating three commits:
1. Rename the dts nodes for 6120
2. create DTS trigger and remove line from 01_leds
3. add R6080

alexclewontin replied:
> > wlan1 uses the MAC address two above it (e.g. B0:39:56:00:00:02, which appears in the factory partition at 0x8044).
> 
> Should be 0x8004? If you have a valid address there that matches 5 GHz MAC address on stock FW, you can remove the
> 
> ```
> mtd-mac-address = <&factory 0x4>;
> mtd-mac-address-increment = <(2)>;
> ```
> 
> from wifi@0,0. (If mtd-eeprom is 0x8000, the MAC address will be read from 0x8004 automatically.)

Yes it should be 0x8004; will do.

alexclewontin replied:
Should now be separated as requested. I do not have access to an R6120, so I cannot test.

alexclewontin commented:
> > wlan1 uses the MAC address two above it (e.g. B0:39:56:00:00:02, which appears in the factory partition at 0x8044).
> 
> Should be 0x8004? If you have a valid address there that matches 5 GHz MAC address on stock FW, you can remove the
> 
> ```
> mtd-mac-address = <&factory 0x4>;
> mtd-mac-address-increment = <(2)>;
> ```
> 
> from wifi@0,0. (If mtd-eeprom is 0x8000, the MAC address will be read from 0x8004 automatically.)

Yes it should be 0x8004; will do.

alexclewontin committed:
ramips: add support for Netgear R6080

This adds support for the Netgear R6080, aka Netgear AC1000.

The R6080 has almost the same hardware as the Netgear R6120,
aka Netgear AC1200, but it lacks the USB port, has only 8 MiB flash and
uses a different SERCOMM_HWID.

Specification:

SoC: MediaTek MT7628 (580 MHz)
Flash: 8 MiB
RAM: 64 MiB
Wireless: 2.4Ghz (builtin) and 5Ghz (MT7612E)
LAN speed: 10/100
LAN ports: 4
WAN speed: 10/100
WAN ports: 1
UART (57600 8N1) on PCB

Flashing OpenWRT from stock firmware requires nmrpflash. Use an ethernet
cable to connect to LAN port 1 of the R6080, and power the R6080 off.
From the connected workstation, run
`nmrpflash -i eth0 -f openwrt-ramips-mt76x8-netgear_r6080-squashfs-factory.img`,
replacing eth0 with the appropriate interface (can be identified by
running `nmrpflash -L`). Then power on the R6080. After flashing has finished,
power cycle the R6080, and it will boot into OpenWRT. Once OpenWRT has been
installed, subsequent flashes can use the web interface and sysupgrade files.

Signed-off-by: Alex Lewontin <alex.c.lewontin@gmail.com>

alexclewontin committed:
ramips: change Netgear R6120 LED labels

This changes the labels for the LEDs in the Netgear R6120
device-tree file, to provide consistency with other devices, especially
the Netgear R6080.

Signed-off-by: Alex Lewontin <alex.c.lewontin@gmail.com>

alexclewontin committed:
ramips: move Netgear R6120 LED trigger

This moves the trigger for the Netgear R6120's wlan2g_green LED from
base-files/etc/board.d/01_leds to the device-tree file.

I do not have access to a Netgear R6120 (only the very similar Netgear
R6080) so this needs testing.

Signed-off-by: Alex Lewontin <alex.c.lewontin@gmail.com>

alexclewontin commented:
@adrianschmutzler all reviews should be addressed, with PR split into 3 commits.

adrianschmutzler commented:
Please include a shortened version of the flashing instructions into the commit message. A reader of the commit message should be able to flash without reading external source (at least in principle).

adrianschmutzler commented on target/linux/ramips/dts/mt7628an_netgear_r6080.dts:
```diff
@@ -0,0 +1,148 @@
+/dts-v1/;
```
Please add a SPDX license identifier.

adrianschmutzler commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -167,6 +167,26 @@ define Device/mercury_mac1200r-v2
 
+define Device/netgear_r6080
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := 7872k
```
I calculated 7552. You would have 7b0000 ...

alexclewontin replied:
You're right, bad math by me. Fixed.

adrianschmutzler commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -167,6 +167,26 @@ define Device/mercury_mac1200r-v2
+  SERCOMM_HWVER := A001
+  SERCOMM_SWVER := 0x0040
+  IMAGES += factory.img
+  IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE)| append-rootfs | \
```
Add a space between BLOCKSIZE and the pipe symbol (and fix that for r6120 as well, please)

alexclewontin replied:
Done. I left this in its own commit for the R6120 because I wasn't sure where it would best logically fit (given the other purely cosmetic commit has already been merged). Let me know if you want me to squash it into the `ramips: move Netgear R6120 LED trigger` commit.

adrianschmutzler commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -167,6 +167,26 @@ define Device/mercury_mac1200r-v2
+  IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE)| append-rootfs | \
+	pad-rootfs
+  IMAGE/sysupgrade.bin := $$(IMAGE/default) | append-metadata | \
+	check-size $$$$(IMAGE_SIZE)
```
checks-size works without argument now.

adrianschmutzler commented on target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds:
```diff
@@ -40,6 +40,7 @@ hiwifi,hc5761a)
 mediatek,linkit-smart-7688)
 	ucidef_set_led_wlan "wifi" "wifi" "linkit-smart-7688:orange:wifi" "phy0tpt"
 	;;
+netgear,r6080|\
```
Since you have a trigger in DTS now, you have to create a separate case for r6080 (you may join them later again when you remove the wifi for r6120).

alexclewontin replied:
Created separate case in the R6080 commit, and then merged them again in the R6120 commit, after the R6120 triggers get moved.

adrianschmutzler commented:
@Brother-Lal @blocktrron 
Does one of you (still) own an R6120 and is able to verify whether the DTS trigger (trivially) introduced here works? It would also be interesting whether the SPI speed could be raised for R6120 (set to 10 MHz there).

adrianschmutzler commented:
@alexclewontin I've pushed the second patch to master already, as this one is just cosmetic.

When doing that, I found that your author field "From:" does not match you Signed-off-by line. You can fix this by `git commit --amend --author "Alex Lewontin <alex.c.lewontin@gmail.com>"`
Despite, Signed-off for your first patch lacks brackets for the mail address.

alexclewontin committed:
ramips: clean up Netgear R6120 code formatting

This commit performs minor janitorial work to clean up some code
formatting for the Netgear R6120.

Signed-off-by: Alex Lewontin <alex.c.lewontin@gmail.com>

alexclewontin commented:
@adrianschmutzler Addressed your comments (including author field on the other commits.) Let me know if the commit adding a space between BLOCKSIZE and the pipe symbol for r6120 should be squashed into another.

adrianschmutzler commented:
@alexclewontin I've merged the remaining patches for R6120 and updated your patch for R6080 to support kernel 5.4 (current default in master for mt76x8).

You may find a commit ready for merge here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

Please review and test on your device. If your response is positive, I will merge it.

alexclewontin commented:
@adrianschmutzler I can confirm that a completely fresh build from the staging branch of your tree ([adrian.git/staging](https://git.openwrt.org/openwrt/staging/adrian.git)) with kernel version 5.4.48 is working perfectly on my R6080.

adrianschmutzler commented:
Thanks, will be merged during the next hour.
